arangod/Pregel/Algos/AIR/Greenspun/Primitives.cpp

```diff
@@ -0,0 +1,1264 @@
+
+EvalResult Prim_AttribRef(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  if (!params.isArray() || params.length() != 2) {
+    return EvalError("expected exactly two parameters");
+  }
+
+  auto&& [slice, key] = unpackTuple<VPackSlice, VPackSlice>(params);
+  auto res = ReadAttribute(slice, key);
+  if (res.fail()) {
+    return res.error();
+  }
+
+  result.add(res.value());
+  return {};
+}
+
+EvalResult Prim_AttribRefOr(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  if (!params.isArray() || params.length() != 3) {
+    return EvalError("expected exactly two parameters");
+  }
+
+  auto&& [slice, key, defaultValue] =
+      unpackTuple<VPackSlice, VPackSlice, VPackSlice>(params);
+  if (!slice.isObject()) {
+    return EvalError("expect second parameter to be an object");
+  }
+
+  auto res = ReadAttribute(slice, key);
+  if (res.fail()) {
+    return res.error();
+  }
+
+  VPackSlice resultValue = res.value();
+  if (resultValue.isNone()) {
+    resultValue = defaultValue;
+  }
+
+  result.add(resultValue);
+  return {};
+}
+
+EvalResult Prim_AttribRefOrFail(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  if (!params.isArray() || params.length() != 2) {
+    return EvalError("expected exactly two parameters");
+  }
+
+  auto&& [slice, key] = unpackTuple<VPackSlice, VPackSlice>(params);
+  if (!slice.isObject()) {
+    return EvalError("expect second parameter to be an object");
+  }
+
+  auto res = ReadAttribute(slice, key);
+  if (res.fail()) {
+    return res.error();
+  }
+
+  VPackSlice resultValue = res.value();
+  if (resultValue.isNone()) {
+    return EvalError("key " + key.toJson() + " not present");
+  }
+
+  result.add(resultValue);
+  return {};
+}
+
+EvalResult Prim_AttribSet(Machine& ctx, VPackSlice const params, VPackBuilder& result) {
+  if (!params.isArray() && params.length() != 3) {
+    return EvalError("expected exactly three parameters");
+  }
+
+  auto&& obj = params.at(0);
+  auto&& key = params.at(1);
+  auto&& val = params.at(2);
+
+  if (!obj.isObject()) {
+    return EvalError("expect first parameter to be an object");
+  }
+
+  if (!key.isString() && !key.isArray()) {
+    return EvalError("expect second parameter to be an array or string");
+  }
+
+  if (key.isString()) {
+    VPackBuilder tmp;
+    {
+      VPackObjectBuilder ob(&tmp);
+      tmp.add(key.copyString(), val);
+    }
+    MergeObjectSlice(result, obj, tmp.slice());
+  } else if (key.isArray()) {
+    VPackBuilder tmp;
+    std::vector<VPackStringRef> path;
+    uint64_t length = key.length();
+    uint64_t iterationStep = 0;
+
+    tmp.openObject();
+    for (auto&& pathStep : VPackArrayIterator(key)) {
+      if (!pathStep.isString()) {
+        return EvalError("expected string in key arrays");
+      }
+      if (iterationStep < (length - 1)) {
+        tmp.add(pathStep.copyString(), VPackValue(VPackValueType::Object));
+      } else {
+        tmp.add(pathStep.copyString(), val);
```

Comment:
This copies the key strings `pathStep.copyString()` unnecessarily.
There is also an `add(...)` overload that does not require an `std::string` for the attribute name.